src/hooks/useTransactionList.ts

```diff
@@ -43,9 +43,38 @@ export function useTransactionList(wallet: EdgeCurrencyWallet, tokenId: string |
     // Sends the mutable state to React,
     // merging the two transaction lists together:
     function requestRender() {
+      // Merge changedTxs into the main txs array.
+      for (const [txid, changedTx] of changedTxs.entries()) {
+        // Use binary search on the existing txs array to find insert location:
+        let left = 0
+        let right = txs.length - 1
+        let insertIndex = 0
+        let replaceCount = 0
+        while (left <= right) {
+          // Divide the sum of left and right by two in order to find the middle
+          // of the binary search (use bitwise operation to avoid floats).
+          const index = (left + right) >> 1
+          const searchTx = txs[index]
+          if (changedTx.date < searchTx.date) {
+            left = index + 1
+          } else if (changedTx.date > searchTx.date) {
+            right = index - 1
+          } else if (changedTx.txid === searchTx.txid) {
+            // Handle the common case where the changedTx is not new
+            replaceCount = 1
+          }
+          // Keep updating the insert location as we search
+          insertIndex = index
+        }
+        // Insert changed transaction
+        txs.splice(insertIndex, replaceCount, changedTx)
+        // Remove it from the changedTxs map because we've committed to txs
+        changedTxs.delete(txid)
+      }
```

Comment:
Can we factor this into a standalone utility at the end of the file, outside the hook? Something like `function binarySearch(txs: EdgeTransaction[], tx: EdgeTransaction): number { ... }`.

Then the logic would be:

```ts
const location = binarySearch(txs, changedTx)
const replaceCount = txs[location].txid === changedTx.txid ? 1 : 0
txs.splice(insertIndex, replaceCount, changedTx)
```

I think this would be more readable.

Alternatively, we could just leverage the set operations, like `new Set([...txs, ...changeTxs])` to deduplicate everything, then do an `Array.sort` based on the date to put it all back in order. The whole algorithm would be like 4 lines long, but definitely not as slick or as fast as what you have:

```ts
const noDupes = txs.filter(tx => changedTxs.has(tx.txid))
noDupes.push(...changedTxs)
return { 
  transactions: noDupes.sort((a, b) => a.date - b.date)
} 
```

Reply:
I woke up with this thought to make it it's own utility.